x-pack/legacy/plugins/integrations_manager/server/integrations.ts

```diff
@@ -0,0 +1,231 @@
+  const searchObjects: SavedObjectsBulkGetObject[] = registryItems.map(({ name, version }) => ({
+    type: SAVED_OBJECT_TYPE,
+    id: `${name}-${version}`,
+  }));
+  const client = getClient(req);
+  const results = await client.bulkGet(searchObjects);
+  const savedObjects: InstallationSavedObject[] = results.saved_objects.filter(o => !o.error); // ignore errors for now
+  const integrationList: IntegrationList = createIntegrationList(registryItems, savedObjects);
+
+  return integrationList;
+}
+
+export async function handleGetInfo(req: PackageRequest) {
+  const { pkgkey } = req.params;
+  const item = await Registry.fetchInfo(pkgkey);
+  const savedObject = await getInstallationObject(getClient(req), pkgkey);
+  const installation = createInstallationObject(item, savedObject);
+
+  return installation;
+}
+
+export async function handleRequestInstall(req: InstallFeatureRequest) {
+  const { pkgkey, feature } = req.params;
+
+  if (feature === 'dashboard') {
+    const toBeSavedObjects = await getObjects(pkgkey, feature);
+    const client = getClient(req);
+    const createResults = await client.bulkCreate(toBeSavedObjects, { overwrite: true });
+    const installed = createResults.saved_objects.map(({ id, type }) => ({ id, type }));
+    const mgrResults = await client.create(
+      SAVED_OBJECT_TYPE,
+      { installed },
+      { id: pkgkey, overwrite: true }
+    );
+
+    return mgrResults;
+  }
+
+  return {
+    pkgkey,
+    feature,
+    created: [],
+  };
+}
+
+export async function handleRequestDelete(req: InstallFeatureRequest) {
+  const { pkgkey, feature } = req.params;
+  const client = getClient(req);
+
+  const installation = await getInstallationObject(client, pkgkey);
+  const installedObjects = (installation && installation.attributes.installed) || [];
+
+  // Delete the manager saved object with references to the asset objects
+  // could also update with [] or some other state
+  await client.delete(SAVED_OBJECT_TYPE, pkgkey);
+
+  // ASK: should the manager uninstall the assets it installed
+  // or just the references in SAVED_OBJECT_TYPE?
+  if (feature === 'dashboard') {
+    // Delete the installed assets
+    const deletePromises = installedObjects.map(async ({ id, type }) => client.delete(type, id));
+    await Promise.all(deletePromises);
+  }
+
+  return {
+    pkgkey,
+    feature,
+    deleted: installedObjects,
+  };
+}
+
+export async function getObjects(pkgkey: string, type: string): Promise<SavedObject[]> {
+  const paths = await Registry.getArchiveInfo(`${pkgkey}.tar.gz`);
+  const toBeSavedObjects = new Map();
+
+  for (const path of paths) {
+    collectReferences(path, toBeSavedObjects, 'dashboard');
+  }
+
+  return Array.from(toBeSavedObjects.values(), ensureJsonValues);
+}
+
+function getAsset(key: string) {
+  const value = cacheGet(key);
+  if (value !== undefined) {
+    const json = value.toString('utf8');
+    return JSON.parse(json);
+  }
+}
+
+function collectReferences(
```

Comment:
We need to support optional assets (e.g. no ML jobs, or only install data files) and reading the saved objects (from the archive) is the best (only?) way to get an asset’s dependency tree so we know what to skip.

At least that’s how I imagine it and how I avoided installing the ingest pipeline while installing all the dashboard-related objects.

Reply:
OK, makes sense for the most part. Can you adjust `collectReferences` to return a `Map` rather than mutate a passed in one? I think that might be a little easier to follow.

Reply:
Updated in d6873c10901deb06f6abad3213590b1bdaac2dd1